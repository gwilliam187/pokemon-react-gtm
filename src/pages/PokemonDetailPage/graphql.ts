import { gql } from "@apollo/client";

export const GET_POKEMON = gql`
  query pokemon($name: String!) {
    pokemon(name: $name) {
      id
      name
      sprites {
        front_default
      }
      moves {
        move {
          url
          name
        }
      }
      types {
        type {
          url
          name
        }
      }
    }
  }
`;

export type TGetPokemonReq = {
  name: string;
};

export type TGetPokemonRes = {
  pokemon: TPokemon;
};

export type TPokemon = {
  id: string;
  name: string;
  sprites: {
    front_default: string;
  };
  moves: TMove[];
  types: TType[];
};

export type TMove = {
  move: TBaseName;
};

export type TType = {
  type: TBaseName;
};

export type TBaseName = {
  url: string;
  name: string;
};
