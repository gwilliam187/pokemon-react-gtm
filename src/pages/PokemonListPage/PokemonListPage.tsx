/** @jsxImportSource @emotion/react */
import { useCallback, useContext, useEffect, useRef, useState } from "react";
import { Link } from "react-router-dom";
import { useLazyQuery } from "@apollo/client";
import { css } from "@emotion/react";

import CircularProgress from "components/CircularProgress";
import FixedBottomNav from "components/FixedBottomNav";
import AppContext from "AppContext";

import {
  GET_POKEMON_LIST,
  TGetPokemonListReq,
  TGetPokemonListRes,
  TPokemonListItem,
} from "./graphql";

const PokemonListPage = () => {
  const { ownedPokemons } = useContext(AppContext);
  const [pokemonList, setPokemonList] = useState<TPokemonListItem[]>([]);

  const loadingRef = useRef<any>(null);
  const [getPokemon, { loading }] = useLazyQuery<
    TGetPokemonListRes,
    TGetPokemonListReq
  >(GET_POKEMON_LIST, {
    onCompleted: (data) => {
      setPokemonList([...pokemonList, ...data.pokemons.results]);
    },
  });

  const handleObserverInView = useCallback(
    (entries: any[]) => {
      if (entries[0].isIntersecting) {
        getPokemon({
          variables: {
            limit: LIMIT,
            offset: pokemonList.length,
          },
        });
      }
    },
    [pokemonList]
  );

  useEffect(() => {
    const observer = new IntersectionObserver(handleObserverInView, {
      root: null,
      rootMargin: "0px",
      threshold: 1.0,
    });

    const loadingRefCurr = loadingRef.current;

    loadingRefCurr && observer.observe(loadingRefCurr);

    return () => {
      observer.disconnect();
    };
  }, [handleObserverInView]);

  return (
    <div
      css={css`
        padding: 1rem;
        margin-bottom: 50px;
      `}
    >
      <h1>All Pokemons</h1>
      <div
        css={css`
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(8rem, 1fr));
          grid-gap: 0.5rem;
        `}
      >
        {pokemonList.map((pokemon) => (
          <Link
            to={`/pokemon/${pokemon.name}`}
            style={{ textDecoration: "none" }}
            key={pokemon.id}
          >
            <div
              css={css`
                display: flex;
                flex-direction: column;
                background-color: #ffffff;
                border: 1px solid #dcdcdc;
                border-radius: 4px;

                & img {
                }

                & p {
                  color: #404040;
                  text-align: center;

                  &:first-letter {
                    text-transform: capitalize;
                  }
                }
              `}
            >
              <img src={pokemon.image} alt={pokemon.name} />
              <p>{pokemon.name}</p>
            </div>
          </Link>
        ))}
      </div>
      <div ref={loadingRef}>
        {loading && (
          <div
            css={css`
              display: flex;
              align-items: center;
              justify-content: center;
              margin-top: 1rem;
            `}
          >
            <CircularProgress />
            <span style={{ marginLeft: "0.5rem" }}>Loading Pokemons</span>
          </div>
        )}
      </div>
      <FixedBottomNav>
        <p>
          <b>Owned:</b> {Object.keys(ownedPokemons).length}{" "}
        </p>
        <Link
          to="/my-pokemon"
          css={css`
            margin-left: auto;
            text-decoration: none;
          `}
        >
          My Pokemons
        </Link>
      </FixedBottomNav>
    </div>
  );
};

export default PokemonListPage;

const LIMIT = 30;
