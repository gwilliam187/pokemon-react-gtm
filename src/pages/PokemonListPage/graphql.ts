import { gql } from "@apollo/client";

export const GET_POKEMON_LIST = gql`
  query pokemons($limit: Int, $offset: Int) {
    pokemons(limit: $limit, offset: $offset) {
      results {
        id
        name
        image
      }
    }
  }
`;

export type TGetPokemonListReq = {
  limit: number;
  offset: number;
};

export type TGetPokemonListRes = {
  pokemons: {
    results: TPokemonListItem[];
  };
};

export type TPokemonListItem = {
  id: string;
  name: string;
  image: string;
};
