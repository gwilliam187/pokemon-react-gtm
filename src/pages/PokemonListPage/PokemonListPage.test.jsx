import { BrowserRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";

import PokemonListPage from "./PokemonListPage";
import { GET_POKEMON_LIST } from "./graphql";
import { act } from "react-dom/test-utils";

describe("<PokemonListPage />", () => {
  it("matches snapshot", async () => {
    window.IntersectionObserver = jest.fn(() => ({
      observe: jest.fn(),
      unobserve: jest.fn(),
      root: jest.fn(),
      rootMargin: jest.fn(),
      thresholds: jest.fn(),
      disconnect: jest.fn(),
      takeRecords: jest.fn(),
    }));

    const MOCKS = [
      {
        request: {
          query: GET_POKEMON_LIST,
          variables: {
            limit: 30,
            offset: 0,
          },
        },
        result: {
          data: {
            pokemons: {
              results: [
                {
                  id: "1",
                  name: "pokemon-1",
                  image: "pokemon-1-image",
                },
              ],
            },
          },
        },
      },
    ];

    act(() => {
      const { container } = render(
        <MockedProvider mocks={MOCKS} addTypename={false}>
          <BrowserRouter>
            <PokemonListPage />
          </BrowserRouter>
        </MockedProvider>
      );

      expect(container).toMatchSnapshot();
    });
  });
});
