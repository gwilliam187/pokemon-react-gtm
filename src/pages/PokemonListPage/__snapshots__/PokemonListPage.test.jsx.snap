// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonListPage /> matches snapshot 1`] = `
<div>
  <div
    class="css-1ltwh5q-PokemonListPage"
  >
    <h1>
      All Pokemons
    </h1>
    <div
      class="css-1gxjgue-PokemonListPage"
    />
    <div />
    <div
      class="css-np6ll6-FixedBottomNav"
    >
      <p>
        <b>
          Owned:
        </b>
         
        0
         
      </p>
      <a
        class="css-1a2wrx3-PokemonListPage"
        href="/my-pokemon"
      >
        My Pokemons
      </a>
    </div>
  </div>
</div>
`;

exports[`<PokemonListPage /> renders list of pokemon according to the GraphQL response 1`] = `
<div>
  <div
    class="css-1ltwh5q-PokemonListPage"
  >
    <h1>
      All Pokemons
    </h1>
    <div
      class="css-1gxjgue-PokemonListPage"
    />
    <div />
    <div
      class="css-np6ll6-FixedBottomNav"
    >
      <p>
        <b>
          Owned:
        </b>
         
        0
         
      </p>
      <a
        class="css-1a2wrx3-PokemonListPage"
        href="/my-pokemon"
      >
        My Pokemons
      </a>
    </div>
  </div>
</div>
`;
